pkg/controllers/user/globaldns/user_cluster_globaldns_controller.go

```diff
@@ -0,0 +1,114 @@
+package globaldns
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/rancher/rancher/pkg/namespace"
+	v1beta1Rancher "github.com/rancher/types/apis/extensions/v1beta1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type UserGlobalDNSController struct {
+	ingresses             v1beta1Rancher.IngressInterface
+	ingressLister         v1beta1Rancher.IngressLister
+	globalDNSs            v3.GlobalDNSInterface
+	globalDNSLister       v3.GlobalDNSLister
+	multiclusterappLister v3.MultiClusterAppLister
+	clusterName           string
+}
+
+func newUserGlobalDNSController(clusterContext *config.UserContext) *UserGlobalDNSController {
+	g := UserGlobalDNSController{
+		ingresses:             clusterContext.Extensions.Ingresses(""),
+		ingressLister:         clusterContext.Extensions.Ingresses("").Controller().Lister(),
+		globalDNSs:            clusterContext.Management.Management.GlobalDNSs(""),
+		globalDNSLister:       clusterContext.Management.Management.GlobalDNSs("").Controller().Lister(),
+		multiclusterappLister: clusterContext.Management.Management.MultiClusterApps("").Controller().Lister(),
+		clusterName:           clusterContext.ClusterName,
+	}
+	return &g
+}
+
+func (g *UserGlobalDNSController) sync(key string, obj *v3.GlobalDNS) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	// This controller only deals with GlobalDNS created with MultiClusterAppID set
+	if obj.Spec.MultiClusterAppName == "" {
+		return nil, nil
+	}
+
+	// If multiclusterappID is set, look for ingresses in the projects of multiclusterapp's targets
+	// Get multiclusterapp by name set on GlobalDNS spec
+	split := strings.SplitN(obj.Spec.MultiClusterAppName, ":", 2)
+	if len(split) != 2 {
```

Comment:
can we extract the code for fetching multi cluster app name from global dns to its own method? I believe we use this code in several places, so need to check them all

Reply:
sure